docs/spec.md

```diff
@@ -593,23 +593,34 @@ Performs element-wise addition of two tensors `lhs` and `rhs` and produces a
 * For integers: integer addition.
 * For floats: `addition` from IEEE-754.
 * For complex numbers: complex addition.
+* For quantized types:
+  * `float_result = (lhs - zero_point(lhs)) * scale(lhs) +
+    (rhs - zero_point(rhs)) * scale(rhs)`.
+  * `rounded_result = round_nearest_even(float_result / scale(result))`.
+  * `result = clamp(storage_min(result), rounded_result + zero_point(result), storage_max(result))`.
 
 #### Inputs
 
-| Label | Name  | Type   | Constraints |
-|-------|-------|--------|-------------|
-| (I1)  | `lhs` | tensor | (C1)        |
-| (I2)  | `rhs` | tensor | (C1)        |
+| Label | Name  | Type                       | Constraints |
+|-------|-------|----------------------------|-------------|
+| (I1)  | `lhs` | tensor or quantized tensor | (C1-C4)     |
+| (I2)  | `rhs` | tensor or quantized tensor | (C1-C3)     |
 
 #### Outputs
 
-| Name     | Type   | Constraints |
-|----------|--------|-------------|
-| `result` | tensor | (C1)        |
+| Name     | Type                       | Constraints |
+|----------|----------------------------|-------------|
+| `result` | tensor or quantized tensor | (C1-C3)     |
 
 #### Constraints
 
-* (C1) `lhs`, `rhs` and `result` have the same type.
+* (C1) `shape(lhs) = shape(rhs) = shape(result)`.
+* If the operation uses non-quantized tensors:
+  * (C2) `element_type(lhs) = element_type(rhs) = element_type(result)`.
+* If the operation uses quantized tensors:
+  * (C3) `element_type(lhs) = element_type(rhs) = element_type(result)`,
+    except for quantization parameters which may differ.
```

Comment:
I think quantization_dimension may also differ because quantization_dimension(lhs) is enforced to be empty.
Also, when they differ, quantization_dimension(result) = quantization_dimension(rhs) will hold.

Reply:
I think in this spec **quantization parameters** only refer to `scales` and `zero_points`.

Xref.

> Quantized element types represent integer values of a storage type in the range from storage_min to storage_max (inclusive) that correspond to floating-point values of an expressed type. For a given integer value `i`, the corresponding floating-point value f can be computed as `f = (i - zero_point) * scale`, where `scale` and `zero_point` are called quantization parameters.

> In quantized tensors, quantization can be per-tensor, meaning, having one `scale` and `zero_point` for the entire tensor or can be per-axis, meaning, having multiple `scales` and `zero_points`, one pair per slice of a particular dimension `quantized_dimension`. More formally, in a tensor t of with per-axis quantization, there are `dim(t, quantized_dimension)` slices of the `quantized_dimension`: `t[:, ..., 0, ..., :]`, `t[:, ..., 1, ..., :]`, etc. All elements in the ith slice use `scales[i]` and `zero_points[i]` as their quantization parameters.

Also, I think it will be necessary to require `quantized_dimension(rhs) == quantized_dimension(result)` (same as current text).  If we allow different `rhs`/`result` `quantized_dimension`, the implementation will need 2-D scales to rescale the immediate result to the output scale.

Reply:
>> I think in this spec quantization parameters only refer to scales and zero_points.
That is right.

>> quantization_dimension(lhs) is enforced to be empty.
This constraint is no longer there in favor of the generality of the semantics. Just an FYI about the updated spec of the add in [here](https://github.com/openxla/stablehlo/blob/main/docs/spec.md#add). The only constraint we have right now is:
`(C1) baseline_type(lhs) = baseline_type(rhs) = baseline_type(result)` which allows only the quantization parameters (scales/zero-points) of `lhs`, `rhs`, and `result` can differ. 

I will get back on the suggestion to differ `quantized_dimension`.

Reply:
@loganchien One follow up question:

If an implementation follows the current way defining semantics `dequant_op_quant`, how does it matter if the `quantization_dimenstion(rhs) != quantization_dimension(result)`. The intermediate floating point results can be quantized back to the result type with any quantization dimension appropriate to the result type. Or am I missing something?

I have updated the https://github.com/openxla/stablehlo/issues/1574 with the above information which addressees per-axis quantization. Let us talk about this more at #1574.

Reply:
If `quantization_dimenstion(rhs) != quantization_dimension(result)`, the scale will become 2-D.  For example:

```
a = dense<[[1, 2], [3, 4], [5, 6]]> : tensor<3x2x!quant.uniform<u8:f32, 0.5:0>>
b = dense<[[1, 2], [3, 4], [5, 6]]> : tensor<3x2x!quant.uniform<u8:f32:0, {3.0:0,4.0:0,5.0:0}>>
c = add(a, b) : (tensor<3x2x!quant.uniform<u8:f32, 0.5:0>>, tensor<3x2x!quant.uniform<u8:f32:0, {3.0:0,4.0:0,5:0}>>) -> tensor<3x2x!quant.uniform<u8:f32:1, {6.0:0,7.0:0}>>
```

The computation for `c`:

```
c[0][0] = (a[0][0] * 0.5 + b[0][0] * 3.0) * 6.0
c[0][1] = (a[0][1] * 0.5 + b[0][1] * 3.0) * 7.0
c[1][0] = (a[1][0] * 0.5 + b[1][0] * 4.0) * 6.0
c[1][1] = (a[1][1] * 0.5 + b[1][1] * 4.0) * 7.0
c[2][0] = (a[1][0] * 0.5 + b[1][0] * 5.0) * 6.0
c[2][1] = (a[1][1] * 0.5 + b[1][1] * 5.0) * 7.0
```

There are 6 scales `3 * 6`, `3 * 7`, `4 * 6`, `4 * 7`, `5 * 6`, and `5 * 7`.